Neighborly/MemoryMappedList.cs

```diff
@@ -0,0 +1,428 @@
+using System.Collections;
+using System.IO.MemoryMappedFiles;
+
+namespace Neighborly;
+
+public class MemoryMappedList : IDisposable, IEnumerable<Vector>
+{
+    private const int s_idBytesLength = 16;
+    private const int s_offsetBytesLength = sizeof(long);
+    private const int s_lengthBytesLength = sizeof(int);
+    private const int s_indexEntryByteLength = s_idBytesLength + s_offsetBytesLength + s_lengthBytesLength;
+    private static readonly Guid s_tombStone;
+    private static readonly byte[] s_tombStoneBytes;
+    private readonly MemoryMappedFileHolder _indexFile;
+    private readonly MemoryMappedFileHolder _dataFile;
+    private readonly object _mutex = new();
+    private long _count;
+    private bool _disposedValue;
+
+    static MemoryMappedList()
+    {
+        s_tombStone = Guid.NewGuid();
+
+        Span<byte> tombStoneBytes = stackalloc byte[16];
+        if (!s_tombStone.TryWriteBytes(tombStoneBytes))
+        {
+            throw new InvalidOperationException("Failed to write the tombstone to bytes");
+        }
+
+        s_tombStoneBytes = tombStoneBytes.ToArray();
+    }
+
+    public MemoryMappedList(long capacity)
+    {
+        _indexFile = new(s_indexEntryByteLength * capacity);
+        _dataFile = new(1024L * capacity);
```

Comment:
It's a bit difficult to find a good maximum here, as the vector size will could differ quite a lot. In my test (running in a devcontainer in WSL2), a capacity of `int.MaxValue * 8L` with 1024 bytes for each item was okay for single tests.

Reply:
To be safe, I recommend a ceiling of 4K. As of now, the typical Vector dimensions are 512, 1024, 768, 2048, and 4096. This depends on the vectorizer and how it chunks.